core/src/storage/seed_encryption.py

```diff
@@ -0,0 +1,217 @@
+            f"Ciphertext length ({len(ciphertext)}) does not match expected length ({current_length})"
+        )
+
+
+def get_decryption_data(session: cache_codec.DataCache) -> Tuple[bytes, bytes]:
+    """
+    Retrieves the decryption data - nonce and tag - from the session attributes.
+    These attributes are set during encryption and are used to decrypt the session's seeds.
+    If the session does not contain encryption data, it raises an error.
+
+    :raises ValueError: If the session does not contain encryption data.
+    """
+    try:
+        tag: bytes = getattr(session, "encryption_tag")
+        nonce: bytes = getattr(session, "encryption_nonce")
+    except AttributeError:
+        raise ValueError("No encryption data found for the session")
+    return nonce, tag
+
+
+def delete_decryption_data(session: cache_codec.DataCache) -> None:
+    """
+    Deletes the decryption data (nonce and tag) from the session attributes.
+    This is used to free memory after decryption is done.
+    If the session does not contain encryption data, it raises an error.
+
+    :raises ValueError: If the session does not contain encryption data.
+    """
+    import gc
+
+    try:
+        delattr(session, "encryption_nonce")
+        delattr(session, "encryption_tag")
+        gc.collect()  # Force garbage collection to free memory
+    except AttributeError:
+        raise ValueError("No encryption data found for the session")
+
+
+def set_decryption_data(
+    session: cache_codec.DataCache, nonce: bytes, tag: bytes
+) -> None:
+    """
+    Sets the decryption data (nonce and tag) as attributes of the session.
+    These are later used to decrypt the session's seeds.
+    Raises an error if the session already contains encryption data.
+
+    :param session: The session to set the decryption data into.
+    :param nonce: The 12-bytes nonce used for encryption.
+    :param tag: The 16-bytes tag used for authentication.
+
+    :raises ValueError: If the session already contains encryption data.
+    """
+    if hasattr(session, "encryption_nonce") or hasattr(session, "encryption_tag"):
+        raise ValueError("Session already contains encryption data")
+    setattr(session, "encryption_nonce", nonce)
+    setattr(session, "encryption_tag", tag)
+
+
+def get_common_data(session: cache_codec.DataCache) -> Tuple[bytes, List[int]]:
+    """
+    Retrieves the session ID and the fields to encrypt based on the DataCache type.
+    For `SessionCache`, it uses the session ID of the session.
+    For `SessionlessCache`, it uses a zeroed session ID.
+
+    :return: the session ID and the list of fields to encrypt/decrypt
+    based on the DataCache type.
+    """
+
+    if isinstance(session, cache_codec.SessionCache):
+        session_id = session.export_session_id()
+        fields_to_encrypt = FIELDS_TO_ENCRYPT
+    elif isinstance(session, cache_common.SessionlessCache):
+        session_id = b"\x00" * cache_codec.SESSION_ID_LENGTH
```

Comment:
Why not override `export_session_id()` in `SessionlessCache` to return an empty string?